### DISCUSSION TITLE
Implementing multi table drop (#2036)

### DISCUSSION THREAD

**Discussion by ghost** [crate/crate#11655] [Q&A]
Hi,
I am starting to learn about crate db and to dive in the code. 
I would like to work on https://github.com/crate/crate/issues/2036 as an exercise and looking from some info/documentation on how to get started. 

I couldn't find an example to follow on how to deal with multi arguments so I started by adding another class : 
```
public final class AnalyzedDropTableMulti implements DDLStatement {

    private final List<AnalyzedDropTable> analyzedDropTables;

    AnalyzedDropTableMulti(@Nullable List<AnalyzedDropTable> analyzedDropTables) {
        this.analyzedDropTables = analyzedDropTables;
    }

    public List<AnalyzedDropTable> getAnalyzedDropTables() {
        return analyzedDropTables;
    }

    @Override
    public <C, R> R accept(AnalyzedStatementVisitor<C, R> visitor, C context) {
        return visitor.visitDropTable(this, context);
    }

    @Override
    public void visitSymbols(Consumer<? super Symbol> consumer) {
    }
}
```
Then changed the DropTableAnalyzer to loop through the table list : 

```
public AnalyzedDropTableMulti analyze(DropTable<?> node, SessionContext sessionContext) {
        List<AnalyzedDropTable> analyzedDropTables = new ArrayList<>();
        for (Table nodeTable : node.tables()) {
            analyzedDropTables.add(analyze(nodeTable.getName(), node.dropIfExists(), sessionContext));
        }
        return new AnalyzedDropTableMulti(analyzedDropTables);
    }

    public AnalyzedDropTableMulti analyze(DropBlobTable<?> node, SessionContext sessionContext) {
        List<AnalyzedDropTable> analyzedDropTables = new ArrayList<>();
        for (Table nodeTable : node.tables()) {
            List<String> parts = nodeTable.getName().getParts();
            if (parts.size() != 1 && !parts.get(0).equals(BlobSchemaInfo.NAME)) {
                throw new IllegalArgumentException("No blob tables in schema `" + parts.get(0) + "`");
            } else {
                QualifiedName name = new QualifiedName(
                    List.of(BlobSchemaInfo.NAME, nodeTable.getName().getSuffix()));
                analyzedDropTables.add(analyze(name, node.ignoreNonExistentTable(), sessionContext));
            }
        }
        return new AnalyzedDropTableMulti(analyzedDropTables);
    }
```

I am not sure how to edit the planner to reflect these changes. 
Can you tell me if this is the expected change or is there a more sophisticated approach?

Thanks,

**Comment by mfussenegger** [comment]
Hi. Nice to hear that you're looking into this.

On a high level, processing a statement involves a couple of steps:

1. Parsing the statement
2. Analysis of the statement
3. Planning/optimizing the statement
4. Executing it.


## For the parser:

For `DROP TABLE` to support dropping more than one table at a time you'd first have to adapt the SQL grammar:

https://github.com/crate/crate/blob/af5bb5acd98aabaadbeee87f508e52d7060f6015/libs/sql-parser/src/main/antlr/SqlBase.g4#L87-L88

Adjust the logic that creates the `DropTable` AST node:

https://github.com/crate/crate/blob/af5bb5acd98aabaadbeee87f508e52d7060f6015/libs/sql-parser/src/main/java/io/crate/sql/parser/AstBuilder.java#L569-L572

You could consider to adjust the existing node to hold multiple `Table<?>` items instead of specializing the many-tables case.

And extend the tests:

https://github.com/crate/crate/blob/af5bb5acd98aabaadbeee87f508e52d7060f6015/libs/sql-parser/src/test/java/io/crate/sql/parser/TestStatementBuilder.java#L234-L239



## For the analyzer:

The analyzer usually turns a `Statement` or `Expression` class of the parsed statement into an analyzed form which contains additional data - like type information, and it also does some basic semantic validation.

For the drop table case, the entry point is here:

https://github.com/crate/crate/blob/af5bb5acd98aabaadbeee87f508e52d7060f6015/server/src/main/java/io/crate/analyze/Analyzer.java#L446-L449


Currently the result of the `DropTableAnalyzer` is a `AnalyzedDropTable`. I think that would still fit - you could adjust the `AnalyzedDropTable` to hold a list of tables instead of a single one.


## Planner / Optimizer

This currently doesn't do much for drop table and can probably be left unchanged. See

https://github.com/crate/crate/blob/af5bb5acd98aabaadbeee87f508e52d7060f6015/server/src/main/java/io/crate/planner/Planner.java#L314-L317


It just wraps the analyzed result into a plan

## Execution

A simple implementation would change the current execution logic to loop over the tables, and execute the same logic we have per table:

https://github.com/crate/crate/blob/af5bb5acd98aabaadbeee87f508e52d7060f6015/server/src/main/java/io/crate/planner/node/ddl/DropTablePlan.java#L74-L116


I think that may be good enough for `DROP TABLE`.

In case of a failure it may happen that one table got deleted, but another didn't. To avoid that we'd also have to adjust the underlying cluster state change operation, but that could also be a follow up task.
